src/paymasters/SignaturePaymaster.sol

```diff
@@ -0,0 +1,144 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.18;
+
+import "@aa/core/BasePaymaster.sol";
+import "@aa/core/UserOperationLib.sol";
+import "@aa/core/UserOperationLib.sol";
+import "@openzeppelin/contracts/utils/Address.sol";
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+
+/**
+ * @notice
+ * A signature paymaster that uses external service to decide whether to pay for the UserOp.
+ * The paymaster trusts an external signer to sign the transaction.
+ * The calling user must pass the UserOp to that external signer first, which performs
+ * whatever off-chain verification before signing the UserOp.
+ * Note that this signature is NOT a replacement for the account-specific signature:
+ * - the paymaster checks a signature to agree to PAY for GAS.
+ * - the account checks a signature to prove identity and account ownership.
+ */
+contract SignaturePaymaster is BasePaymaster, Initializable, UUPSUpgradeable {
+    using UserOperationLib for UserOperation;
+    using ECDSA for bytes32;
+
+    address public immutable verifyingSigner;
+
+    uint256 private constant VALID_TIMESTAMP_OFFSET = 20;
+    uint256 private constant SIGNATURE_OFFSET = 84;
+
+    mapping(address => uint256) public senderNonce;
+
+    // ========== Constructor & Upgrades ============
+
+    constructor(IEntryPoint _entryPoint, address _verifyingSigner) BasePaymaster(_entryPoint) {
+        _disableInitializers();
+        verifyingSigner = _verifyingSigner;
+    }
+
+    /**
+     * @dev The _entryPoint member is immutable, to reduce gas consumption.  To upgrade EntryPoint,
+     * a new implementation of SignaturePaymaster must be deployed with the new EntryPoint address.
+     */
+    function initialize(address owner) external virtual initializer {
+        __UUPSUpgradeable_init();
+        _transferOwnership(owner);
+    }
+
+    /**
+     * @dev Authorize the upgrade. Only by an owner.
+     * @param newImplementation address of the new implementation
+     */
+    // This function is called by the proxy contract when the implementation is upgraded
+    function _authorizeUpgrade(address newImplementation) internal view override {
+        require(msg.sender == owner(), "VP: not owner");
+        (newImplementation);
+    }
+
+    /* =============== Viewers & Validation ============= */
+
+    function pack(UserOperation calldata userOp) internal pure returns (bytes memory ret) {
+        // lighter signature scheme. must match UserOp.ts#packUserOp
+        bytes calldata pnd = userOp.paymasterAndData;
+        // copy directly the userOp from calldata up to (but not including) the paymasterAndData.
+        // this encoding depends on the ABI encoding of calldata, but is much lighter to copy
+        // than referencing each field separately.
+        assembly {
+            let ofs := userOp
```

Comment:
🧙 assembly, I cannot review this on my head, assume it is correct xD

Reply:
lol, i've just checked it out and looks alright, but would be good to have tests as well, just in case

Reply:
it's exactly the same as the one from Argent though, right? i guess they have it audited

Reply:
Yes. It is taken from audited code, so we should be fine. Also, I understand it and it looks correct to me.